File: 2 Add Two Numbers/C/listnode.c
#include <stdio.h>
#include <stdlib.h>

#include "listnode.h"

struct ListNode* createNode(int val) {
    struct ListNode* newNode = NULL;
    newNode = (struct ListNode*) malloc(sizeof(struct ListNode));
    
    if (newNode == NULL) {
        perror("Cannot allocate memory for a new node!");
        exit(1);
    }

    newNode->val = val;
    newNode->next = NULL;

    return newNode;
}

struct ListNode* append(struct ListNode* head, int val) {
    struct ListNode* newNode = NULL;
    newNode = createNode(val);

    if (head == NULL) {
        return newNode;
    }

    struct ListNode* temp = head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = newNode;
    return head;
}

void print(struct ListNode* head) {
    printf("[");
    while (head->next != NULL) {
        printf("%d, ", head->val);
        head = head->next;
    }
    printf("%d]\n", head->val);
}

void freeList(struct ListNode* head) {
    struct ListNode* current = head;
    struct ListNode* next = NULL;

    while (current != NULL) {
        next = current->next;
        free(current);
        current = next;
    }
}

File: 2 Add Two Numbers/Java/ListNode.java
public class ListNode {
    int val;
    ListNode next;

    ListNode() {}

    ListNode(int val) { 
        this.val = val; 
    }

    ListNode(int val, ListNode next) { 
        this.val = val; 
        this.next = next; 
    }

    public static void print(ListNode head) {
        System.out.print("[");
        while (head.next != null) {
            System.out.print(head.val + ", ");
            head = head.next;
        }
        System.out.print(head.val + "]\n");
    }
}
